Improve some ListenetSet gep descriptions (#3978)
kubernetes-sigs/gateway-api: geps/gep-1713/index.md

}

type AllowedListeners struct {

> 	// +kubebuilder:default={from: None}

Reviewer: I may be out of context, wdym here? are we moving away from the Same as default?

	// 2. ListenerSet ordered by creation time (oldest first)
	// 3. ListenerSet ordered alphabetically by “{namespace}/{name}”.
	//

> 	// Regarding Conflict Management, Listeners in a ListenerSet follow the same

Reviewer 2: Should we add this new clause to the API types godoc?

Also we might want to call out sectionName behaves a bit differently where sibling ListenerSets can have the same sectionName.

Author: so, per Nick's suggestion I have used a similar approach of https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/gateway_types.go#L72 that contains the full description on godoc.

Reviewer 3: Does this contradict the GEP ?

https://gateway-api.sigs.k8s.io/reference/spec/#gatewayspec
The implementation MUST NOT pick one conflicting Listener as the winner.
ALL indistinct Listeners must not be accepted for processing.

https://gateway-api.sigs.k8s.io/geps/gep-1713/#listener-validation
Implementations MUST treat the parent Gateways as having the merged list of all listeners from itself and attached ListenerSets and validation of this list of listeners MUST behave the same as if the list were part of a single Gateway with the relaxed listener name constraints.

Reviewer 3: ```suggestion
	// Regarding Conflict Management, Listeners within in a ListenerSet follow the same
```

Author: I have at least removed it from the CRD generation, closing this for now

	// rules of Listeners on a Gateway resource.
	// 

> 	// This validation should happen within all of the ListenerSets attached to a 

Reviewer 3: I think there should be two sections - (i) validation of listeners in a listenerSet (ii) listenerSets in a gateway with examples
Reading this, I would think that all listeners should be validated together as is done in a Gateway

> 	// This validation should happen within all of the ListenerSets attached to a 

Reviewer 3: ```suggestion
	// This validation should also happen within all of the ListenerSets attached to a 
```
This might disambiguate when the listeners should be validated within the listenerset context and in the gateway context

The following example shows a `Gateway` with an HTTP listener and two child HTTPS `ListenerSets` with unique hostnames and certificates.

> Only `ListenerSets` from the same namespace of the `Gateway` will be accepted:

Author: I have removed the double line break, but kept at least one between 343 and 344 to avoid having a huge amount of columns on a single line (easier to review!)

    sectionName: second
```

> To attach to listeners in both a `Gateway` and `ListenerSet` the route MUST have two `parentRefs`:

Reviewer 2: Reading this now it is a duplicate of `Routes MUST be able to attach to a `ListenerSet` and it's parent `Gateway` by having multiple `parentRefs` eg:`

I saw we replace the above yaml

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: httproute-example
spec:
  parentRefs:
  - name: second-workload-listeners
    kind: ListenerSet
    sectionName: second
```

with the yaml containing two parent refs

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: httproute-example
spec:
  parentRefs:
  - name: second-workload-listeners
    kind: ListenerSet
    sectionName: second
  - name: parent-gateway
    kind: Gateway
    sectionName: foo
```

For instance, the following `HTTPRoute` attempts to attach to a listener defined in the parent `Gateway` using the sectionName `foo`. This is not valid and the route's status `Accepted` condition should be set to `False`

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: httproute-example
spec:
  parentRefs:

>   - name: some-workload-listeners

Reviewer: I guess no real change here, just order, right?

Author: right, I wrote on my self review why I did the change. The example of the attachment of Gateway AND ListenerSet is more clear on this order, then we can go to the invalid examples :)

Reviewer 2: To make this invalid the `kind` needs to be a `ListenerSet` in the parentRef

    kind: Gateway
    sectionName: foo
```

> >--- Ricardo - above is a bit confusing, maybe ask Dave some clarification

Reviewer: /cc @dprotaso

> >--- Ricardo - above is a bit confusing, maybe ask Dave some clarification

Author: @dprotaso the example above is a bit confusing for me, probably because the API changed from when it was implemented.

I was wondering what it means for an HTTPRoute to try to attach a listener defined on a Gateway, using a section name `foo`. If this section name exists on the Gateway it is valid, right? eg.:

```
  listeners:
  - name: foo
    port: 80
    protocol: HTTP
```

So how can the situation above be invalid? Maybe we can get the full yaml definition, with Gateway, ListenerSet and HTTPRoute for clarification here?

Thanks!

spec:
  parentRefs:
  - name: second-workload-listeners
    kind: XListenerSet
    sectionName: second
```

> If routes MUST attach to a `XListenerSet` and its parent `Gateway`, it MUST have multiple `parentRefs` eg:

Reviewer 4: Nit: "MUST" feels out of place here. This is meant to be used for instructions for implementers. Instead I'd reword this to

```suggestion
To attach a Route to a `XListenerSet` and its parent `Gateway`, it MUST have multiple `parentRefs` eg:
```

Author: fixed

  name: httproute-example
spec:
  parentRefs:
  - name: second-workload-listeners
    kind: ListenerSet
    sectionName: second
  - name: parent-gateway
    kind: Gateway
    sectionName: foo
```

For instance, the following `HTTPRoute` attempts to attach to a listener defined in the parent `Gateway` using the sectionName `foo`, which also exists on a ListenerSet.

> This is not valid and the route's status `Accepted` condition should be set to `False`

Reviewer 5: Actually, that HTTPRoute should, as written, fail to attach, because the `kind` is `Gateway`, not `ListenerSet` (and that really should be, for now, `XListenerSet`, here and everywhere else).

Assuming that the Kind is corrected, then I actually don't think this should be an error on the HTTPRoute, because there is only one Listener it could attach to. 

Now that we've clarified that there's no Listener attachment across parentRefs, then we can make the "can this attach" rule a bit simpler by saying "When a Route tries to attach to a Listener using a `sectionName`, then that `sectionName` must be distinct _within that Listener group (which could be a Gateway or a ListenerSet)." I think the situation shouldn't arise because `listener.name` should also be distinct within a Listener group, but it's good to be specific, I think.

> This is not valid and the route's status `Accepted` condition should be set to `False`

Author: so, echoing as a manifest because I understand yaml better than people :P 

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: parent-gateway
spec:
  gatewayClassName: example
  allowedListeners:
    namespaces:
      from: Same
  listeners:
  - name: foo
    hostname: foo.com
    protocol: HTTP
    port: 80
  - name: foo
    hostname: foo1.com
    protocol: HTTP
    port: 80
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: httproute-example
spec:
  parentRefs:
  - name: parent-gateway
    kind: Gateway
    sectionName: foo
```

But isn't this the case already? Don't we have a conflict management already in place for it, that does not allow two liteners on a listenergroup to have the same name?

> This is not valid and the route's status `Accepted` condition should be set to `False`

Reviewer 5: Yes, that's what I meant by saying "this situation should not arise".

It's probably good to have the behavior specified anyway, because I think it clarifies the overall intent better.

> This is not valid and the route's status `Accepted` condition should be set to `False`

Reviewer 5: But yes, that YAML should not work, because `name` can't be repeated across Listeners.

> This is not valid and the route's status `Accepted` condition should be set to `False`

Reviewer 3: >  "if a ListenerSet has a sectionName, and gateway not, but you say the parentRef is a gateway and try to use the ListenerSet sectionName this should fail"

This makes it clear, thanks!

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: parent-gateway
spec:
  gatewayClassName: example
  allowedListeners:
    namespaces:
      from: Same
  listeners:
  - name: foo2
    hostname: foo.com
    protocol: HTTP
    port: 80
  - name: foo3
    hostname: foo1.com
    protocol: HTTP
    port: 80
---
apiVersion: gateway.networking.x-k8s.io/v1alpha1
kind: ListenerSet
metadata:
  name: first-workload-listeners
spec:
  parentRef:
    name: parent-gateway
    kind: Gateway
    group: gateway.networking.k8s.io
  listeners:
  - name: foo
    hostname: first.foo.com
    protocol: HTTP
    port: 80
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: httproute-example
spec:
  parentRefs:
  # No SectionName is set on this parentRef
  - name: some-workload-listeners 
    kind: Gateway
```

The example above SHOULD be attached only to the following listeners:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
.....
spec:
  listeners:
  - name: foo
    hostname: foo.com
    protocol: HTTP
    port: 80
  - name: foo1
    hostname: foo1.com
    protocol: HTTP
    port: 80

> ```

Author: oops, thanks!

  name: httproute-example
spec:
  parentRefs:
  - name: second-workload-listeners
    kind: XListenerSet
    sectionName: second
  - name: parent-gateway
    kind: Gateway
    sectionName: foo
```

For instance, the following `HTTPRoute` attempts to attach to a listener defined in the parent `Gateway` using the sectionName `foo`, which exists on a `ListenerSet` but not on a `Gateway`.

> This is not valid and the route's status `Accepted` condition should be set to `False`

Reviewer 4: What should a Gateway implementation do with this? In the case where a Route is no longer accepted, is it supposed to drop it altogether? IE if you add a new parentRef that makes the Route invalid, do you drop the entire Route, or does the Gateway impl add something into status specifically for this ListenerSet?

Actually, did we cover how status is supposed to work here? Should we populate separate Route parent statuses for each ListenerSet and Gateway, or just the Gateway? cc @youngnick

Author: oh, sorry, I just typed and saw your answer @davidjumani 

So, the current behavior today is that the route is not dropped, and it gets something like invalid Parent ref on the status?

Reviewer 5: In the case highlighted here, if a HTTPRoute references a `sectionName` on a `parentRef` that points to a ListenerSet, and there is no Listener in that ListenerSet with that `name` field set, then the HTTPRoute MUST fail to attach, regardless of what Listeners exist at the Gateway level.

That's what we agreed on earlier - to summarize:

* For considering Listener-Listener conflicts, you have to look at all Listeners, regardless of their source.
* For considering Route attachement, you only need to consider the listeners in the actual parent object (Gateway or ListenerSet). For Routes that attach to a Gateway, you don't need to care about what Listeners are in any ListenerSets, and for Routes that attach to a ListenerSet, you don't need to care about what Listeners are in the Gateway.

Maybe another way to say this that's more specification-like:

* Routes attach to all **relevant** Listeners in its **direct Parent** object.
* If a Route has a parentRef that is a ListenerSet, then the ListenerSet is the **direct Parent**.
* If a Route has a parentRef that is a Gateway, then the Gateway is the **direct Parent**.
* If the Route is using Gateway defaulting, then the default Gateway is the **direct Parent**.
* If `sectionName` is not set, then all Listeners in the direct Parent object are **relevant**, and the Route must attach to all of them (subject to any `allowedRoutes` criteria.)
* If `sectionName` is set, then only the Listener in the direct Parent object with the `name` field set to the same value as `sectionname` is **relevant**, and the Route must attach only to that Listener (subject to any `allowedRoutes` criteria).
* If there are no relevant Listeners (because, for example, a sectionName is specified that does not exist on the direct Parent), then the Route has nowhere to attach to and MUST be have `Accepted` set to `false`.

Reviewer 4: That's a phenomenal summary, thanks @youngnick! Would love to see this incorporated into the GEP.

Author: @robscott @youngnick I have copied Nick's text as is to the GEP, as I agree it is great and very clarifying, thanks!

2. ListenerSet ordered by creation time (oldest first)
3. ListenerSet ordered alphabetically by “{namespace}/{name}”.

Conflicts are covered in the section [Listener and ListenerSet conflicts](#listener-and-listenerset-conflicts)

### Listener and ListenerSet conflicts

ListenerSet conflicts should be managed similarly to [Gateway resource conflict](https://github.com/kubernetes-sigs/gateway-api/blob/372a5b06624cff12117f41dcd26c08cb1def22e7/apis/v1/gateway_types.go#L76)
management.

With ListenerSet this validation should happen within the same ListenerSet resource, 
but MUST be validated also within a Gateway scope and all of the attached Listeners/ListenerSets.

This means that the validation should happen now between distinct ListenerSets 
attached to the same Gateway, and in case of a conflict, the [Listener Precedence](#listener-precedence) 
should be respected, and the conflicting listener MUST have a `Conflicted` condition 

> set to True and with an explicit reason on its message.

Author: I don't think so. I think from a conflict management perspective, the most precedent Listener is marked as Accepted, the others are conflicted (as it is done on Gateway Listeners)

> set to True and with an explicit reason on its message.

Reviewer 3: Thanks, makes it much clearer!

> set to True and with an explicit reason on its message.

Reviewer 3: This iteration of the doc explains things clearly. Thanks!

2. ListenerSet ordered by creation time (oldest first)
3. ListenerSet ordered alphabetically by “{namespace}/{name}”.

Conflicts are covered in the section [Listener and ListenerSet conflicts](#listener-and-listenerset-conflicts)

> ### Listener and ListenerSet conflicts

Reviewer 2: We should combine this section into `ListenerConditions within a ListenerSet` so that conflicts are covered in one section.

Author: IMHO having a different section that defines what a conflict means from "what are the conditions" make it more clear for implementations.

We can probably refer on "ListenerConditions within a ListenerSet" back to the `### Listener and ListenerSet conflicts` to define what a conflict means, but I think that adding all on a single "Conditions" section would overload it and make it confusing.

Reviewer 2: > IMHO having a different section that defines what a conflict means from "what are the conditions" make it more clear for implementations.

Sure sounds good

ListenerSet conflicts should be managed similarly to [Gateway resource conflict](https://github.com/kubernetes-sigs/gateway-api/blob/372a5b06624cff12117f41dcd26c08cb1def22e7/apis/v1/gateway_types.go#L76)
management.

With ListenerSet this validation should happen within the same ListenerSet resource, 

> but MUST be validated also within a Gateway scope and all of the attached Listeners/ListenerSets.

Reviewer 2: this language needs to exclude sectionName when comparing it across sibling Listeners

Author: "With ListenerSet this validation should happen within the same ListenerSet resource, but MUST be validated also within a Gateway scope and all of the attached Listeners/ListenerSets. The SectionName field is an exception for this validation, and while it should not conflict within the same ListenerSet, it can be duplicated between different ListenerSets" ?

This means that the validation should happen now between distinct ListenerSets 
attached to the same Gateway, and in case of a conflict, the [Listener Precedence](#listener-precedence) 
should be respected, and the conflicting listener MUST have a `Conflicted` condition 
set to True and with an explicit reason on its message.

Following are some examples of a conflict situation:

#### Conflict between ListenerSet and parent Gateway

Given the following resource definitions:

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: parent-gateway
  namespace: infra
spec:
  allowedListeners:
    namespaces:
      from: All
  listeners:
  - name: foo
    hostname: www.something.tld
    protocol: HTTPS
    port: 443

> 	tls:

Reviewer 2: tab instead of space maybe?

<img width="187" height="84" alt="Image" src="https://github.com/user-attachments/assets/d93a8e71-005c-49fb-99dc-929f92778caa" />

      mode: Terminate
      certificateRefs:
      - kind: Secret
        group: ""
        name: default-cert
---
apiVersion: gateway.networking.x-k8s.io/v1alpha1
kind: ListenerSet
metadata:
  name: user-listenerset
  namespace: user01
spec:
  parentRef:
    name: parent-gateway
    kind: Gateway
    group: gateway.networking.k8s.io
  listeners:
  - name: myapp
    hostname: www.something.tld
    protocol: HTTPS
    port: 443
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        group: ""
        name: app-cert
```

The ListenerSet `user-listenerset` should be marked as Conflicted, as the `parent-gateway`
have a listener definition called `foo` that conflicts with the ListenetSet definition
called `myapp`, as the following:

```yaml
apiVersion: gateway.networking.x-k8s.io/v1alpha1
kind: ListenerSet
metadata:
  name: user-listenerset
  namespace: user01
....
status:
  listeners:
  - name: myapp
    hostname: www.something.tld
    protocol: HTTPS
    port: 443

> 	conditions:

Reviewer 2: using a tab here instead of space

Author: fixed, thanks!

>     - message: ListenerSet has conflicts with Gateway 'infra/parent-gateway'

Reviewer 2: Would we want to highlight the listener name in the error message?

Author: I didn't wanted to leak cross namespace resource names

Reviewer 2: sounds good

2. ListenerSet ordered by creation time (oldest first)
3. ListenerSet ordered alphabetically by “{namespace}/{name}”.

Conflicts are covered in the section [Listener and ListenerSet conflicts](#listener-and-listenerset-conflicts)

### Listener and ListenerSet conflicts

ListenerSet conflicts should be managed similarly to [Gateway resource conflict](https://github.com/kubernetes-sigs/gateway-api/blob/372a5b06624cff12117f41dcd26c08cb1def22e7/apis/v1/gateway_types.go#L76)
management.

With ListenerSet this validation should happen within the same ListenerSet resource, 
but MUST be validated also within a Gateway scope and all of the attached Listeners/ListenerSets. The SectionName field is an exception for this validation, and while 
it should not conflict within the same ListenerSet, it can be duplicated between 

> different ListenerSets.

Reviewer 4: Nit: The line length here feels inconsistent

Reviewer 6: (FTR I think the the Mesh GEP experiment was worth doing, but I think that the sentence-per-line approach is likely to be more practical.)
